File: 351004/Halukha/publisher/src/main/java/com/example/lab/publisher/service/MarkerService.java
package com.example.lab.publisher.service;

import java.util.List;
import java.util.stream.Collectors;

import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.CachePut;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.cache.annotation.Caching;
import org.springframework.stereotype.Service;

import com.example.lab.publisher.dto.MarkerRequestTo;
import com.example.lab.publisher.dto.MarkerResponseTo;
import com.example.lab.publisher.exception.EntityNotFoundException;
import com.example.lab.publisher.mapper.MarkerMapper;
import com.example.lab.publisher.model.Marker;
import com.example.lab.publisher.repository.MarkerRepository;

@Service
public class MarkerService {

    private final MarkerRepository markerRepository;
    private final MarkerMapper mapper = MarkerMapper.INSTANCE;

    public MarkerService(MarkerRepository newsRepository) {
        this.markerRepository = newsRepository;
    }

    @Cacheable(cacheNames = "markers", key = "'all'")
    public List<MarkerResponseTo> getAllMarker() {
        return markerRepository.findAll().stream()
                .map(mapper::toDto)
                .collect(Collectors.toList());
    }

    @Cacheable(cacheNames = "markers", key = "#id")
    public MarkerResponseTo getMarkerById(Long id) {
        return markerRepository.findById(id)
                .map(mapper::toDto)
                .orElseThrow(() -> new EntityNotFoundException("Marker not found", 40401));
    }

    @Caching(evict = {
            @CacheEvict(cacheNames = "markers", key = "'all'")
    })
    public MarkerResponseTo createMarker(MarkerRequestTo request) {
        Marker news = mapper.toEntity(request);
        Marker saved = markerRepository.save(news);
        return mapper.toDto(saved);
    }

    @Caching(put = {
            @CachePut(cacheNames = "markers", key = "#id")
    }, evict = {
            @CacheEvict(cacheNames = "markers", key = "'all'")
    })
    public MarkerResponseTo updateMarker(Long id, MarkerRequestTo request) {
        Marker existing = markerRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Marker not found", 40401));
        Marker updated = mapper.updateEntity(request, existing);
        updated.setId(id);
        Marker saved = markerRepository.save(updated);
        return mapper.toDto(saved);
    }

    @Caching(evict = {
            @CacheEvict(cacheNames = "markers", key = "#id"),
            @CacheEvict(cacheNames = "markers", key = "'all'")
    })
    public void deleteMarker(Long id) {
        if (!markerRepository.existsById(id)) {
            throw new EntityNotFoundException("Marker not found", 40401);
        }
        markerRepository.deleteById(id);
    }
}

File: 351003/brezgunov/distcomp/src/main/kotlin/com/distcomp/service/MarkerService.kt
package com.distcomp.service

import com.distcomp.dto.marker.MarkerRequestTo
import com.distcomp.dto.marker.MarkerResponseTo
import com.distcomp.exception.MarkerNotFoundException
import com.distcomp.mapper.MarkerMapper
import com.distcomp.repository.MarkerRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class MarkerService(
    val markerMapper: MarkerMapper,
    val markerRepository: MarkerRepository
) {
    @Transactional
    fun createMarker(markerRequestTo: MarkerRequestTo): MarkerResponseTo {
        val marker = markerMapper.toMarkerEntity(markerRequestTo)
        markerRepository.save(marker)
        return markerMapper.toMarkerResponse(marker)
    }

    fun readMarkerById(id: Long): MarkerResponseTo {
        val marker = markerRepository.findByIdOrNull(id)
            ?: throw MarkerNotFoundException("Marker with id $id not found")
        return markerMapper.toMarkerResponse(marker)
    }

    fun readAll(): List<MarkerResponseTo> {
        return markerRepository.findAll().map { markerMapper.toMarkerResponse(it) }
    }

    @Transactional
    fun updateMarker(markerRequestTo: MarkerRequestTo, markerId: Long?): MarkerResponseTo {
        if (markerId == null || markerRepository.findByIdOrNull(markerId) == null) {
            throw MarkerNotFoundException("Marker with id $markerId not found")
        }

        val marker = markerMapper.toMarkerEntity(markerRequestTo)
        marker.id = markerId

        markerRepository.save(marker)
        return markerMapper.toMarkerResponse(marker)
    }

    @Transactional
    fun removeMarkerById(id: Long) {
        if (markerRepository.findByIdOrNull(id) == null) {
            throw MarkerNotFoundException("Marker with id $id not found")
        }
        markerRepository.deleteById(id)
    }
}